crates/bevy_tasks/src/task_pool.rs

```diff
@@ -104,10 +162,102 @@ impl TaskPool {
         TaskPoolBuilder::new().build()
     }
 
+    fn any_panicking_threads(&self) -> bool {
+        self.inner
+            .read()
+            .threads
+            .iter()
+            .any(|state| state.panicking())
+    }
+
+    pub(crate) fn handle_panicking_threads(&self) {
+        match self.panic_policy {
+            ThreadPanicPolicy::Propagate => {
+                if self.any_panicking_threads() {
+                    for state in self.inner.write().threads.drain(..) {
+                        let thread = state.thread().clone();
+                        if let Err(err) = state.handle.join() {
+                            panic!(
+                                "TaskPool's inner thread '{:?}' panicked with error: {:?}",
+                                thread, err
+                            );
+                        }
+                    }
+                }
+            }
+            ThreadPanicPolicy::Restart => {
+                if self.any_panicking_threads() {
+                    for (idx, state) in self
+                        .inner
+                        .write()
+                        .threads
+                        .iter_mut()
+                        .filter(|state| state.panicking())
+                        .enumerate()
+                    {
+                        let thread_name = match state.thread().name() {
+                            Some(name) => name.to_owned(),
+                            None => format!("TaskPool ({})", idx),
+                        };
+
+                        let old_state = std::mem::replace(
+                            state,
+                            Self::spawn_thread_internal(
+                                None,
+                                thread_name,
+                                self.executor.clone(),
+                                self.shutdown_rx.clone(),
+                            ),
+                        );
+
+                        // join the panicked thread handle
+                        let panic_error = old_state.handle.join().unwrap_err();
```

Comment:
I'm curious if people have a strong opinion on this behavior. 
Currently a new thread is spawned _before_ the old on is joined. 
Do we want to join the old thread before and then spawn the new one or is this current behavior ok?